### PR TITLE
abstract the parquet coalescing reading

### PR DESCRIPTION
**Pull request by wbo4958**
This PR is to abstract the common coalescing reading logic and apply it to Parquet.

### CODE REVIEW

**Comment by tgravescs** [comment]
please add description

**Comment by tgravescs** [comment]
@wbo4958 sorry I didn't get to review this and forgot about it, can you bring it up to date and I'll review?

**Comment by wbo4958** [comment]
@tgravescs I am refining this PR, will bring it up when it can be reviewable. thx Tom.

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 1]
```diff
@@ -963,387 +985,233 @@ class MultiFileParquetPartitionReader(
...
+    //  and offsets are larger, check to make sure we allocated enough memory before writing.
+    // Not sure how expensive this is, we could throw exception instead if the written
+    // size comes out > then the estimated size.
+    val actualFooterSize = calculateParquetFooterSize(blocks, clippedSchema)
+    // 4 + 4 is for writing size and the ending PARQUET_MAGIC.
+    val bufferSizeReq = offset + actualFooterSize + 4 + 4
+
+    var buf: HostMemoryBuffer = hmb
+    val totalBufferSize = if (bufferSizeReq > initTotalSize) {
+      logWarning(s"The original estimated size $initTotalSize is to small, " +
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1169]:
```suggestion
      logWarning(s"The original estimated size $initTotalSize is too small, " +
```
?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 2]
```diff
@@ -963,387 +985,233 @@ class MultiFileParquetPartitionReader(
...
-        if (numPartitions > 1) {
-          concatAndAddPartitionColsToBatch(cb, rowsPerPartition, inPartitionValues)
-        } else {
-          // single partition, add like other readers
-          addPartitionValues(Some(cb), inPartitionValues.head, partitionSchema).get
-        }
-      }
-    } else {
-      batch
+  override def getFileFormatShortName: String = "Parquet"
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1108]:
```suggestion
  override final def getFileFormatShortName: String = "Parquet"
```
?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 3]
```diff
@@ -176,12 +180,12 @@ abstract class MultiFileCloudPartitionReaderBase(
...
    * Get ThreadPoolExecutor to run the Callable.
    *
-   * There're two rules:
+   * The requirements:
    * 1. Same ThreadPoolExecutor for cloud and coalescing for the same file format
    * 2. Different file formats have different ThreadPoolExecutors
    *
    * @param numThreads  max number of threads to create
-   * @return A ThreadPoolExecutors to used
+   * @return ThreadPoolExecutors
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 4]
```diff
@@ -963,387 +985,233 @@ class MultiFileParquetPartitionReader(
...
-      newhmb
+    val schemaNextFile =
+      nextBlockInfo.schema.asGroupType().getFields.asScala.map(_.getName)
+    val schemaCurrentfile =
+      currentBlockInfo.schema.asGroupType().getFields.asScala.map(_.getName)
+
+    if (!schemaNextFile.sameElements(schemaCurrentfile)) {
+      logInfo(s"File schema for the next file ${nextBlockInfo.filePath}" +
+        s" doesn't match current ${currentBlockInfo.filePath}, splitting it into another batch!")
+      return true
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1045]: we need to return here

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 5]
```diff
@@ -963,387 +985,233 @@ class MultiFileParquetPartitionReader(
...
+   * @return Boolean
+   */
+  override def checkIfNeededToSplitDataBlock(currentBlockInfo: SingleDataBlockInfo,
+      nextBlockInfo: SingleDataBlockInfo): Boolean = {
+    // We need to ensure all files we are going to combine have the same datetime
+    // rebase mode.
+    if (nextBlockInfo.isCorrectedRebaseMode != currentBlockInfo.isCorrectedRebaseMode) {
+      logInfo(s"datetime rebase mode for the next file ${nextBlockInfo.filePath} is " +
+        s"different then current file ${currentBlockInfo.filePath}, splitting into another batch.")
+      return true
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1034]: the return is necessary here.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 6]
```diff
@@ -302,3 +306,487 @@ abstract class MultiFileCloudPartitionReaderBase(
...
+      val filesAndBlocks = LinkedHashMap[Path, ArrayBuffer[DataBlockBase]]()
+      blocks.foreach { case (path, block) =>
+        filesAndBlocks.getOrElseUpdate(path, new ArrayBuffer[DataBlockBase]) += block
+      }
+      val tasks = new java.util.ArrayList[Future[(Seq[DataBlockBase], Long)]]()
+
+      val allBlocks = blocks.map(_._2)
+
+      // First, estimate the output file size for further allocating a buffer for it,
+      val initTotalSize = calculateEstimatedBlocksOutputSize(allBlocks, clippedSchema)
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:609]: Yeah, good question, we have limited the block size by maxReadBatchSizeBytes which is an integer type, So we don't need to check the overflow

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 1 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 2 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
@tgravescs, Could you help to review this PR

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 7]
```diff
@@ -963,387 +981,186 @@ class MultiFileParquetPartitionReader(
...
-            s"small, actual written: ${amountWritten}")
-        }
-        if (finalizehmb != null) {
-          finalizehmb.close()
-        }
-        (hmb, amountWritten)
-      }
-    }
+  override def getThreadPool(numThreads: Int): ThreadPoolExecutor = {
+    ParquetMultiFileThreadPoolFactory.getThreadPool(getFileFormatShortName, numThreads)
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1052]: so it looks like this is changing to share a threadpool between coalescing and cloud, I assume that means the other can be removed. (MultiFileThreadPoolFactory).

**Comment by tgravescs** [comment]
Also what testing have you done here?  I would like to run a few performance tests to make sure no regressions, if we can run on a bunch of small files and compares current version vs this version that would be great.

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
fyi - still 2 questions/comments open

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 8]
```diff
@@ -963,387 +981,186 @@ class MultiFileParquetPartitionReader(
...
-        allPartCols =>
-          ColumnarPartitionReaderWithPartitionValues.addGpuColumVectorsToBatch(cb, allPartCols)
-      }
+  override final def getFileFormatShortName: String = "Parquet"
+
+  override def readBufferToTable(dataBuffer: HostMemoryBuffer, dataSize: Long,
+      isCorrectRebaseMode: Boolean, clippedSchema: SchemaBase): Table = {
+
+    // Dump parquet data into a file
+    if (debugDumpPrefix != null) {
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Thx, Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 9]
```diff
@@ -302,3 +305,493 @@ abstract class MultiFileCloudPartitionReaderBase(
...
+      val filesAndBlocks = LinkedHashMap[Path, ArrayBuffer[DataBlockBase]]()
+      blocks.foreach { case (path, block) =>
+        filesAndBlocks.getOrElseUpdate(path, new ArrayBuffer[DataBlockBase]) += block
+      }
+      val tasks = new java.util.ArrayList[Future[(Seq[DataBlockBase], Long)]]()
+
+      val allBlocks = blocks.map(_._2)
+
+      // First, estimate the output file size for the initial allocating
+      val initTotalSize = calculateEstimatedBlocksOutputSize(allBlocks, clippedSchema)
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:650]: Done. Thx Tom

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
Hi @tgravescs 

I've tested the below scenarios for `upstream` and `upstream + this PR`

- small file testing

total 4000 small parquet files, total 16M

||upstream + this PR time(s)| upstream time(s)|
|---|----|---|
| 1st |  35.736 |33.976|
|2nd|27.406|27.346|
|3rd|23.98|23.442|
|4th|26.217|26.331|

- a litter bigger file testing

total 100 parquet files,  each file is 14M, total 1.3G

||upstream + this PR, time(s)| upstream time(s) |
|---|----|---|
| 1st |  13.836  | 13.421|
|2nd|12.348|12.289|
|3rd|11.676|11.175|
|4th|10.966|10.842|

Seems this PR does not bring performance regression.

**Review by tgravescs — APPROVED** [review]

**Pull request merged**